src/sage/combinat/tuple.py

```diff
@@ -90,6 +90,88 @@ def __repr__(self):
         """
         return "Tuples of %s of length %s" % (self.S, self.k)
 
+    def unrank(self, i):
+        """
+        Return the `i`-th element of the set of Tuples.
+
+        INPUT:
+
+        - ``i`` -- integer between `0` and `n-1` where `n` is the cardinality
+          of this set.
+
+        EXAMPLES::
+
+            sage: T = Tuples(range(7), 6)
+            sage: T[73451]
+            (0, 0, 1, 4, 2, 4)
+
+        TESTS:
+
+        Verify that `unrank` is giving the correct result. ::
+
+            sage: T = Tuples(range(4), 5)
+            sage: all(T[i] == x for i,x in enumerate(T))
+            True
+
+        Verify that `unrank` is fast for large inputs. ::
+
+            sage: Tuples(range(3), 30)[10^12]
+            (1, 0, 0, 1, 1, 1, 2, 0, 1, 2, 0, 1, 1, 0, 2, 1, 1, 0, 1, 2, 1, 2,
+            1, 1, 0, 1, 0, 0, 0, 0)
+
+        Verify that `unrank` normalizes ``i``. ::
+
+            sage: T = Tuples(range(3), 2)
+            sage: T[QQ(4/2)]
+            (2, 0)
+            sage: T[QQ(1/2)]
+            Traceback (most recent call last):
+            ...
+            TypeError: no conversion of this rational to integer
+
+        Verify that `unrank` throws an error when ``i`` is out of bounds. ::
+
+            sage: T = Tuples(range(3), 3)
+            sage: T[27]
+            Traceback (most recent call last):
+            ...
+            IndexError: index i (=27) is greater than or equal to the cardinality
+            sage: T[-28]
+            Traceback (most recent call last):
+            ...
+            IndexError: i (=-1) must be a nonnegative integer
+
+        Verify that `unrank` works correctly for Tuples where `k = 1`. ::
+
+            sage: T = Tuples(range(6), 1)
+            sage: T[5]
+            (5,)
+
+        Verify that `unrank` works when called directly. ::
+
+            sage: T = Tuples(range(4), 3)
+            sage: T.unrank(19)
+            (3, 0, 1)
+
+        Verify that :issue:`39534` has been fixed. ::
+
+            sage: Tuples(range(3), 30).random_element() #random
+            (0, 2, 2, 1, 2, 1, 0, 2, 0, 2, 1, 0, 0, 2, 1, 1, 2, 0, 2, 1, 1, 0,
+            2, 2, 0, 0, 0, 2, 1, 1)
```

Comment:
```suggestion
            sage: T = Tuples(range(3), 30).random_element()
            sage: all(v in range(3) for v in T)
            True
            sage: len(T)
            30
```
Better to test the properties of the random element. `#random` just checks there is no failure.